File: apps/web/lib/github-integration.ts
'use client';

import { z } from 'zod';

const repositorySchema = z.object({
  repository_id: z.string(),
  owner: z.string(),
  name: z.string(),
  full_name: z.string(),
  default_branch: z.string(),
  private: z.boolean(),
  archived: z.boolean(),
  connected: z.boolean().default(false),
});

const connectionSchema = z.object({
  id: z.string(),
  repository_id: z.string(),
  owner: z.string(),
  name: z.string(),
  default_branch: z.string(),
  root_path: z.string(),
  agent_id: z.string(),
  environment_id: z.string(),
  status: z.string(),
  recipe_version: z.string(),
});

const proposedChangeSchema = z.object({
  path: z.string(),
  operation: z.enum(['create', 'update']),
  content_sha: z.string(),
  replacement: z.string(),
  rationale: z.string(),
});

const manualStepSchema = z.object({
  label: z.string(),
  command: z.string(),
  reason: z.string(),
});

const jobSchema = z.object({
  id: z.string(),
  connection_id: z.string(),
  status: z.string(),
  risk_statement: z.string(),
  analysis_summary: z
    .object({
      detected_framework: z.string(),
      package_manager: z.string(),
      summary: z.string(),
      integration_points: z.array(z.string()).default([]),
    })
    .nullable()
    .optional(),
  proposed_changes: z.array(proposedChangeSchema).default([]),
  manual_steps: z.array(manualStepSchema).default([]),
  pull_request_url: z.string().nullable().optional(),
  error_code: z.string().nullable().optional(),
  error_message: z.string().nullable().optional(),
  first_verified_trace_at: z.string().nullable().optional(),
});

export type GitHubRepository = z.infer<typeof repositorySchema>;
export type GitHubConnection = z.infer<typeof connectionSchema>;
export type GitHubIntegrationJob = z.infer<typeof jobSchema>;

export async function createInstallUrl(): Promise<string> {
  const res = await fetch('/api/github-integration/install-url', {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({}),
  });
  const data = await readJson(res);
  return z.object({ install_url: z.string().url() }).parse(data).install_url;
}

export async function listRepositories(): Promise<GitHubRepository[]> {
  const res = await fetch('/api/github-integration/repositories');
  const data = await readJson(res);
  return z.object({ repositories: z.array(repositorySchema) }).parse(data).repositories;
}

export async function listConnections(agentId: string): Promise<GitHubConnection[]> {
  const params = new URLSearchParams({ agent_id: agentId });
  const res = await fetch(`/api/github-integration/connections?${params}`);
  const data = await readJson(res);
  return z.object({ connections: z.array(connectionSchema) }).parse(data).connections;
}

export async function createConnection(input: {
  repositoryId: string;
  rootPath: string;
  agentId: string;
  environmentId: string;
}): Promise<GitHubConnection> {
  const res = await fetch('/api/github-integration/connections', {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({
      repository_id: input.repositoryId,
      root_path: input.rootPath,
      agent_id: input.agentId,
      environment_id: input.environmentId,
    }),
  });
  return connectionSchema.parse(await readJson(res));
}

export async function createJob(input: {
  connectionId: string;
  riskStatement: string;
  sourceProcessingConsent: boolean;
}): Promise<GitHubIntegrationJob> {
  const res = await fetch('/api/github-integration/jobs', {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({
      connection_id: input.connectionId,
      risk_statement: input.riskStatement,
      source_processing_consent: input.sourceProcessingConsent,
    }),
  });
  return jobSchema.parse(await readJson(res));
}

export async function getJob(jobId: string, signal?: AbortSignal): Promise<GitHubIntegrationJob> {
  const init: RequestInit = signal ? { signal } : {};
  const res = await fetch(`/api/github-integration/jobs/${encodeURIComponent(jobId)}`, init);
  return jobSchema.parse(await readJson(res));
}

export async function approveJob(jobId: string): Promise<GitHubIntegrationJob> {
  const res = await fetch(`/api/github-integration/jobs/${encodeURIComponent(jobId)}/approve`, {
    method: 'POST',
  });
  const data = await readJson(res);
  return jobSchema.parse(z.object({ job: jobSchema }).parse(data).job);
}

async function readJson(res: Response): Promise<unknown> {
  const data: unknown = await res.json().catch(() => ({}));
  if (!res.ok) {
    const message =
      data !== null &&
      typeof data === 'object' &&
      'message' in data &&
      typeof data.message === 'string'
        ? data.message
        : data !== null &&
            typeof data === 'object' &&
            'error' in data &&
            typeof data.error === 'string'
          ? data.error
          : 'GitHub integration request failed';
    throw new Error(message);
  }
  return data;
}

File: crates/tl-core/src/github_integration.rs
//! Wire types for GitHub-assisted agent installation.
//!
//! Rust owns the durable installation, repository connection, job lifecycle,
//! proposal, and verification state. Web routes are same-origin adapters only.

use serde::{Deserialize, Serialize};

#[cfg(feature = "schema")]
use schemars::JsonSchema;
#[cfg(feature = "ts-export")]
use ts_rs::TS;
#[cfg(feature = "openapi")]
use utoipa::ToSchema;

pub const GITHUB_INTEGRATION_RECIPE_TYPESCRIPT_NEXTJS_V1: &str = "typescript-nextjs-v1";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub enum GitHubInstallationStatus {
    Active,
    Suspended,
    Removed,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub enum GitHubConnectionStatus {
    Active,
    AccessRemoved,
    Disconnected,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub enum GitHubIntegrationJobStatus {
    Queued,
    Analyzing,
    AwaitingApproval,
    Applying,
    DraftPrOpen,
    AwaitingVerification,
    Verified,
    ClosedUnmerged,
    Error,
    Cancelled,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub enum GitHubRepositorySelection {
    Selected,
    All,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub enum GitHubProposedFileOperation {
    Create,
    Update,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubInstallUrlRequest {
    #[serde(default)]
    #[cfg_attr(feature = "ts-export", ts(optional))]
    pub redirect_path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubInstallUrlResponse {
    pub install_url: String,
    /// RFC 3339 timestamp.
    pub expires_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubCallbackRequest {
    pub code: String,
    pub state: String,
    pub installation_id: String,
    #[serde(default)]
    #[cfg_attr(feature = "ts-export", ts(optional))]
    pub setup_action: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubCallbackResponse {
    pub installation: GitHubInstallationSummary,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubInstallationSummary {
    pub id: String,
    pub workspace_id: String,
    pub installation_id: String,
    pub account_login: String,
    pub account_type: String,
    pub repository_selection: GitHubRepositorySelection,
    pub status: GitHubInstallationStatus,
    /// RFC 3339 timestamp.
    pub created_at: String,
    /// RFC 3339 timestamp.
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubRepositorySummary {
    pub repository_id: String,
    pub owner: String,
    pub name: String,
    pub full_name: String,
    pub default_branch: String,
    pub private: bool,
    pub archived: bool,
    #[serde(default)]
    pub connected: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubRepositoryListResponse {
    pub repositories: Vec<GitHubRepositorySummary>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubConnectionCreateRequest {
    pub repository_id: String,
    pub root_path: String,
    pub agent_id: String,
    pub environment_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubConnectionSummary {
    pub id: String,
    pub workspace_id: String,
    pub installation_id: String,
    pub repository_id: String,
    pub owner: String,
    pub name: String,
    pub default_branch: String,
    pub root_path: String,
    pub agent_id: String,
    pub environment_id: String,
    pub status: GitHubConnectionStatus,
    pub recipe_version: String,
    /// RFC 3339 timestamp.
    pub created_at: String,
    /// RFC 3339 timestamp.
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubConnectionListResponse {
    pub connections: Vec<GitHubConnectionSummary>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubIntegrationJobCreateRequest {
    pub connection_id: String,
    pub risk_statement: String,
    #[serde(default)]
    pub source_processing_consent: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubIntegrationJobSummary {
    pub id: String,
    pub workspace_id: String,
    pub connection_id: String,
    pub status: GitHubIntegrationJobStatus,
    pub risk_statement: String,
    pub base_branch: String,
    #[serde(default)]
    #[cfg_attr(feature = "ts-export", ts(optional))]
    pub base_sha: Option<String>,
    pub recipe_version: String,
    #[serde(default)]
    #[cfg_attr(feature = "ts-export", ts(optional))]
    pub analysis_summary: Option<GitHubIntegrationAnalysisSummary>,
    #[serde(default)]
    pub proposed_changes: Vec<GitHubProposedFileChange>,
    #[serde(default)]
    pub manual_steps: Vec<GitHubIntegrationManualStep>,
    #[serde(default)]
    #[cfg_attr(feature = "ts-export", ts(optional))]
    pub branch_name: Option<String>,
    #[serde(default)]
    #[cfg_attr(feature = "ts-export", ts(optional))]
    pub commit_sha: Option<String>,
    #[serde(default)]
    #[cfg_attr(feature = "ts-export", ts(optional))]
    pub pull_request_number: Option<i64>,
    #[serde(default)]
    #[cfg_attr(feature = "ts-export", ts(optional))]
    pub pull_request_url: Option<String>,
    #[serde(default)]
    #[cfg_attr(feature = "ts-export", ts(optional))]
    pub error_code: Option<String>,
    #[serde(default)]
    #[cfg_attr(feature = "ts-export", ts(optional))]
    pub error_message: Option<String>,
    #[serde(default)]
    #[cfg_attr(feature = "ts-export", ts(optional))]
    pub pr_opened_at: Option<String>,
    #[serde(default)]
    #[cfg_attr(feature = "ts-export", ts(optional))]
    pub pr_merged_at: Option<String>,
    #[serde(default)]
    #[cfg_attr(feature = "ts-export", ts(optional))]
    pub first_verified_trace_at: Option<String>,
    /// RFC 3339 timestamp.
    pub created_at: String,
    /// RFC 3339 timestamp.
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubIntegrationJobListResponse {
    pub jobs: Vec<GitHubIntegrationJobSummary>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubIntegrationAnalysisSummary {
    pub detected_framework: String,
    pub package_manager: String,
    pub summary: String,
    #[serde(default)]
    pub integration_points: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubProposedFileChange {
    pub path: String,
    pub operation: GitHubProposedFileOperation,
    pub content_sha: String,
    pub replacement: String,
    pub rationale: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubIntegrationManualStep {
    pub label: String,
    pub command: String,
    pub reason: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubIntegrationApproveResponse {
    pub job: GitHubIntegrationJobSummary,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(JsonSchema))]
#[cfg_attr(feature = "openapi", derive(ToSchema))]
#[cfg_attr(feature = "ts-export", derive(TS))]
#[cfg_attr(feature = "ts-export", ts(export))]
pub struct GitHubIntegrationCancelResponse {
    pub job: GitHubIntegrationJobSummary,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn status_wire_values_are_snake_case() {
        assert_eq!(
            serde_json::to_string(&GitHubIntegrationJobStatus::AwaitingApproval).unwrap(),
            "\"awaiting_approval\""
        );
        assert_eq!(
            serde_json::to_string(&GitHubInstallationStatus::Active).unwrap(),
            "\"active\""
        );
    }

    #[test]
    fn connection_status_wire_values_are_snake_case() {
        assert_eq!(
            serde_json::to_string(&GitHubConnectionStatus::AccessRemoved).unwrap(),
            "\"access_removed\""
        );
        assert_eq!(
            serde_json::from_str::<GitHubProposedFileOperation>("\"create\"").unwrap(),
            GitHubProposedFileOperation::Create
        );
    }
}
